test/Dialect/FIRRTL/inliner.mlir

```diff
@@ -1267,3 +1267,118 @@ firrtl.module private @test3() {
+        firrtl.matchingconnect %c_i, %i : !firrtl.enum<Some: uint<8>, None: uint<0>>
+        %0 = firrtl.ref.send %c_o : !firrtl.uint<8>
+        %1 = firrtl.ref.cast %0 : (!firrtl.probe<uint<8>>) -> !firrtl.probe<uint<8>, @I::@J>
+        firrtl.ref.define %o, %1 : !firrtl.probe<uint<8>, @I::@J>
+      }
+    }
+  }
+}
+
+// -----
+
+// Test inlining module containing various operations with blocks.
+
+firrtl.circuit "InlineBlocks" {
+  firrtl.layer @I  inline {
+    firrtl.layer @J  inline { }
+  }
+  firrtl.module private @HasBlocks(in %i: !firrtl.enum<Some: uint<8>, None: uint<0>>,
+                                   in %cond: !firrtl.uint<1>,
+                                   out %p: !firrtl.probe<uint<8>, @I::@J>) attributes {annotations = [{class = "firrtl.passes.InlineAnnotation"}]} {
+    firrtl.layerblock @I {
+      firrtl.when %cond : !firrtl.uint<1> {
+        firrtl.layerblock @I::@J {
+          %o = firrtl.wire interesting_name : !firrtl.uint<8>
+          %0 = firrtl.ref.send %o : !firrtl.uint<8>
+          %1 = firrtl.ref.cast %0 : (!firrtl.probe<uint<8>>) -> !firrtl.probe<uint<8>, @I::@J>
+          firrtl.ref.define %p, %1 : !firrtl.probe<uint<8>, @I::@J>
+          firrtl.match %i : !firrtl.enum<Some: uint<8>, None: uint<0>> {
+            case Some(%arg0) {
+              firrtl.matchingconnect %o, %arg0 : !firrtl.uint<8>
+            }
+            case None(%arg0) {
+              %invalid_ui8 = firrtl.invalidvalue : !firrtl.uint<8>
+              firrtl.matchingconnect %o, %invalid_ui8 : !firrtl.uint<8>
+            }
+          }
+          %unused = firrtl.node %o : !firrtl.uint<8>
+        }
+        %unused = firrtl.node %cond : !firrtl.uint<1>
+      }
+    }
+  }
+  // CHECK: @InlineBlocks
+  firrtl.module @InlineBlocks(in %i: !firrtl.enum<Some: uint<8>, None: uint<0>>, in %cond: !firrtl.uint<1>, out %o: !firrtl.probe<uint<8>, @I::@J>) attributes {convention = #firrtl<convention scalarized>} {
+    // Check inlined structure.
+    // CHECK:      layerblock @I
+    // CHECK-NEXT:   firrtl.when
+    // CHECK-NEXT:     firrtl.layerblock @I::@J
+    // CHECK-NEXT:       firrtl.wire
+    // CHECK:            firrtl.match
+    // CHECK:              Some(
+    // CHECK:              None(
+    // CHECK:              }
+    // CHECK-NEXT:       }
+    // CHECK-NEXT:       firrtl.node {{.*}}
+    // CHECK-NEXT:     }
+    // CHECK-NEXT:     firrtl.node {{.*}}
+    // CHECK-NEXT:   }
+    // CHECK-NEXT: }
+    %c_i, %c_cond, %c_p = firrtl.instance c interesting_name @HasBlocks(in i: !firrtl.enum<Some: uint<8>, None: uint<0>>, in cond: !firrtl.uint<1>, out p: !firrtl.probe<uint<8>, @I::@J>)
+    firrtl.matchingconnect %c_i, %i : !firrtl.enum<Some: uint<8>, None: uint<0>>
+    firrtl.matchingconnect %c_cond, %cond : !firrtl.uint<1>
+    firrtl.ref.define %o, %c_p : !firrtl.probe<uint<8>, @I::@J>
+  }
+}
```

Comment:
These tests are a tad large. It may be better to factor this into narrower tests of features. E.g., just inlining into a layerblock of a trivial module. There are also a lot of operations which could be pruned, e.g., invalid connections.